### PR TITLE
Add reduce_agg Presto aggregate function

### PR DESCRIPTION
**Pull request by mbasmanova**
See #6434

### CODE REVIEW

File: velox/docs/functions/presto/aggregate.rst [hunk 1]
```diff
@@ -119,6 +119,44 @@ General Aggregate Functions
...
+
+    Reduces all non-NULL input values into a single value. ``inputFunction``
+    will be invoked for each non-NULL input value. If all inputs are NULL, the
+    result is NULL. In addition to taking the input value, ``inputFunction``
+    takes the current state, initially ``initialState``, and returns the new state.
+    ``combineFunction`` will be invoked to combine two states into a new state.
+    The final state is returned. Throws an error if ``initialState`` is NULL or
+    ``inputFunction`` or ``combineFunction`` returns a NULL.::
+
+        SELECT id, reduce_agg(value, 0, (a, b) -> a + b, (a, b) -> a + b)
```
Comment (pedroerp) [inline velox/docs/functions/presto/aggregate.rst:132]: the difference between inputFunction and combineFunction might not be terribly clear to folks reading this at first. Maybe we can have the second example calculating, say, an average, to give a more illustrative example?

File: velox/exec/Aggregate.h [hunk 2]
```diff
@@ -18,6 +18,7 @@
 #include <folly/Synchronized.h>
 
 #include "velox/common/memory/HashStringAllocator.h"
+#include "velox/core/ExpressionEvaluator.h"
```
Comment (pedroerp) [inline velox/exec/Aggregate.h:21]: Could move setLambdaExpressions() to .cpp and forward declare ExpressionEvaluator to break the header/header dependency.

File: velox/functions/prestosql/aggregates/ReduceAgg.cpp [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+namespace facebook::velox::aggregate::prestosql {
+namespace {
+
+using ReduceAggAccumulator = functions::aggregate::SingleValueAccumulator;
+
+// clang-format off
+//  reduce_agg aggregate function doesn't lend itself well to vectorized
+//  execution as it is effectively a data-dependent loop. It is hard to avoid
+//  evaluating lambda expressions on small number of rows at a time. This
+//  particular implementation doesn't try very hard and simply evaluated lambda
```
Comment (pedroerp) [inline velox/functions/prestosql/aggregates/ReduceAgg.cpp:32]: s/evaluated/evaluates

File: velox/functions/prestosql/aggregates/ReduceAgg.cpp [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+//   s2 = s0
+//   s1 = f(s1, 1)
+//   s2 = f(s2, 10)
+//   s1 = f(s1, 2)
+//   s2 = f(s2, 20)
+//   s1 = f(s1, 3)
+//
+//   The inputFunction lambda expression is evaluated 5 times, once per row.
+//
+//   A more efficient approach would be to evaluate a set of rows that contain
```
Comment (pedroerp) [inline velox/functions/prestosql/aggregates/ReduceAgg.cpp:54]: s/contain/contains

File: velox/functions/prestosql/aggregates/ReduceAgg.cpp [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+  void destroy(folly::Range<char**> groups) override {
```
Comment (pedroerp) [inline velox/functions/prestosql/aggregates/ReduceAgg.cpp:164]: is there anything special about this method? Why isn't this defined in the base class?

**Review by pedroerp — APPROVED** [review]
Looks great to me with a few small comments. Might be worth it to tag the comment with a "TODO" to make it more explicit

File: velox/functions/prestosql/aggregates/ReduceAgg.cpp [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+  }
+
+  void toIntermediate(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      VectorPtr& result) const override {
+    if (!rows.hasSelections()) {
+      // Set all rows in 'result' to null.
+      result->resize(rows.size());
+      memset(
```
Comment (xiaoxmeng) [inline velox/functions/prestosql/aggregates/ReduceAgg.cpp:354]: nit: ::memset

File: velox/docs/functions/presto/aggregate.rst (hunk 1 above)
Comment (mbasmanova) [reply velox/docs/functions/presto/aggregate.rst:132]: It is indeed not straightforward. To compute avg, one needs to compute (sum, count) using reduce_agg first, then compute avg from that as a separate projection. Will add this example to the doc.

```
select sum_and_count.sum / sum_and_count.count from (

  select reduce_agg(x, cast(row(0, 0) as row(sum double, count bigint)), 
    (s, x) -> cast(row(s.sum + x, s.count + 1) as row(sum double, count bigint)), 
    (s, s2) -> cast(row(s.sum + s2.sum, s.count + s2.count) as row(sum double, count bigint))) as sum_and_count
  from unnest(array[1,2,3]) as t(x)

);
```

File: velox/functions/prestosql/aggregates/ReduceAgg.cpp (hunk 5 above)
Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/ReduceAgg.cpp:164]: @pedroerp This method requires knowing the type of the accumulator, which is not available in the base class. Also, some aggregate functions require more logic in 'destroy'.

File: velox/exec/HashAggregation.cpp [hunk 7]
```diff
@@ -140,6 +174,17 @@ HashAggregation::HashAggregation(
         argTypes,
         resultType,
         driverCtx->queryConfig());
+
+    auto lambdas = extractLambdaInputs(aggregate);
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/HashAggregation.cpp:178]: SimpleExpressionEvaluator's constructor doesn't allocate memory.

```
  SimpleExpressionEvaluator(core::QueryCtx* queryCtx, memory::MemoryPool* pool)
      : queryCtx_(queryCtx), pool_(pool) {}
```

File: velox/parse/Expressions.cpp [hunk 8]
```diff
@@ -320,35 +321,60 @@ TypedExprPtr Expressions::resolveLambdaExpr(
...
+}
+
+const exec::FunctionSignature* findLambdaSignature(
+    const std::vector<std::shared_ptr<exec::AggregateFunctionSignature>>&
+        signatures,
+    const std::shared_ptr<const CallExpr>& callExpr) {
+  const exec::FunctionSignature* matchingSignature = nullptr;
+  for (const auto& signature : signatures) {
+    if (isLambdaSignature(signature.get(), callExpr)) {
+      VELOX_CHECK_NULL(
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/parse/Expressions.cpp:361]: This is test code, so it doesn't matter.

**Comment by mbasmanova** [comment]
@pedroerp @xiaoxmeng Pedro, Meng, thank you for reviews. Updated to address comments.

**Pull request closed without merging**